tests/test_queries.py

```diff
@@ -0,0 +1,890 @@
+            ),
+        },
+    }
+
+
+def read_datasets(path):
+    datasets = Datasets(
+        comstock=read_table(path / "data" / "conus_2022_reference_comstock" / "1.0.0"),
+        resstock=read_table(path / "data" / "conus_2022_reference_resstock" / "1.0.0"),
+        tempo=read_table(path / "data" / "tempo_conus_2022" / "1.0.0"),
+    )
+    return datasets
+
+
+def read_table(path):
+    spark = SparkSession.builder.appName("dgrid").getOrCreate()
+    load_data = spark.read.parquet(str(path / "load_data.parquet")).cache()
+    lookup = spark.read.parquet(str(path / "load_data_lookup.parquet")).cache()
+    table = load_data.join(lookup, on="id").drop("id").cache()
+    return Tables(load_data, lookup, table)
+
+
+def perform_op_by_electricity(stats, table, name, operation):
+    if name in ("comstock", "resstock"):
+        columns = ["electricity_cooling", "electricity_heating"]
+    elif name == "tempo":
+        columns = ["L1andL2"]
+    else:
+        assert False, name
+    for col in columns:
+        op = operation.__name__
+        col_name = f"{op}_{col}"
+        if op not in stats[name]:
+            stats[name][op] = {}
+        val = getattr(
+            table.agg(operation(col).alias(col_name)).collect()[0],
+            col_name,
+        )
+        if op == "sum":
+            # 2 scenarios x 2 model years
+            val *= 4
+        stats[name][op][col] = val
+    stats[name]["count"] = table.count()
+
+
+# The next two functions are for ad hoc testing.
+
+
+def run_query(
+    query_name,
+    registry_path=REGISTRY_PATH,
+    operation="sum",
+    output_dir=Path("queries"),
+    persist_intermediate_table=True,
+    load_cached_table=True,
+):
+    setup_logging(
+        "dsgrid", "query.log", console_level=logging.INFO, file_level=logging.INFO, mode="w"
+    )
+    project = Project.load(
+        "dsgrid_conus_2022",
+        offline_mode=True,
+        registry_path=registry_path,
+    )
+    if query_name == QueryTestElectricityValues.NAME:
+        query = QueryTestElectricityValues(registry_path, project, output_dir=output_dir)
+    elif query_name == QueryTestElectricityAgg.NAME:
+        query = QueryTestElectricityAgg(operation, registry_path, project, output_dir=output_dir)
+    elif query_name == QueryTestElectricityUseByCounty.NAME:
+        query = QueryTestElectricityUseByCounty(registry_path, project, output_dir=output_dir)
+    elif query_name == QueryTestElectricityUseByState.NAME:
+        query = QueryTestElectricityUseByState(registry_path, project, output_dir=output_dir)
+    else:
+        raise Exception(f"no query for {query_name}")
+
+    ProjectQuerySubmitter(project, output_dir).submit(
+        query.make_query(),
+        persist_intermediate_table=persist_intermediate_table,
+        load_cached_table=load_cached_table,
+    )
+    result = query.validate()
+    print(f"Result of query {query.name} = {result}")
+
+
+def run_derived_dataset(
```

Comment:
What is a derived_dataset in this context? In the queries dir, there is a derived_dataset folder and a cached_tables folder. What is the difference here?

Reply:
The contents are the same. The difference is that a derived dataset has a dataset_id and can be directly queried. I think it provides a more explicit workflow. Suppose a user wants to do this:
1. Create a dataset that includes comstock, resstock, and tempo. Don't perform any aggregations. This is a long, slow process and the user only wants to run it once.
2. Explore the data.
3. Perform additional queries to answer specific questions.

The user could achieve this by running a project query with no aggregation, then run additional queries with aggregations. It would be fast because of the caching, but they might not understand why. It also wouldn't be obvious on how to view the cached table.